File: Blind 75/11 - Graphs/261-graph-valid-tree.py
'''
*LeetCode premium problem
'''

from time import time
from typing import List


class Solution:
    def isValidTree(self, n: int, edges: List[List[int]]) -> bool:
        # Empty graph counts as a tree
        if len(edges) == 0:
            return True

        node_map = {i: [] for i in range(n)}
        for a, b in edges:
            '''
            Assume edge is directed even though description says undirected
            by tracking and skipping previous node after visiting
            (false positive cycle)
            '''
            node_map[a].append(b)

        seen = set()

        def dfs(curr):
            if curr in seen:
                return False

            seen.add(curr)
            for v in node_map[curr]:
                if not dfs(v):
                    # Cycle detected
                    return False
            return True

        # If there are more components or groups of nodes, len(seen) != n
        return dfs(0) and len(seen) == n

    def reference(self, n: int, edges: List[List[int]]) -> bool:
        if not n:
            return True
        adj = {i: [] for i in range(n)}
        for n1, n2 in edges:
            adj[n1].append(n2)
            adj[n2].append(n1)

        visit = set()

        def dfs(i, prev):
            if i in visit:
                return False

            visit.add(i)
            for j in adj[i]:
                if j == prev:
                    continue
                if not dfs(j, i):
                    return False
            return True

        return dfs(0, -1) and n == len(visit)

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.isValidTree(*case))
                else:
                    self.isValidTree(*case)
        print(f'Runtime for our solution: {time() - sol_start}\n')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(*case))
                else:
                    self.reference(*case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [
        (5, [[0, 1], [0, 2], [0, 3], [1, 4]]),
        # Additional
        (0, []),
        (5, [[0, 1], [0, 2], [0, 3], [1, 4], [0, 4]]),
        (7, [[0, 1], [0, 2], [3, 5], [5, 6], [1, 4]]),
        (7, [[0, 1], [0, 2], [3, 5], [5, 6], [1, 4], [0, 4]]),
    ]
    test.quantify(test_cases)

File: Blind 75/11 - Graphs/JavaScript/261-graph-valid-tree.js
const { performance } = require("perf_hooks");

class Solution {
  method(n, edges) {
    let adjList = Array.from({ length: n }, () => []);
    for (const [a, b] of edges) {
      adjList[a].push(b);
    }

    let seen = new Set();

    const dfs = (curr, prev) => {
      if (seen.has(curr)) return false;

      seen.add(curr);
      for (const n of adjList[curr]) {
        if (n === prev) continue;
        if (!dfs(n, curr)) return false;
      }

      return true;
    };

    return dfs(0, -1) && seen.size === n;
  }

  reference(n, edges) {
    const union = new Array(n).fill(-1);

    for (const [src, dst] of edges) {
      const [x, y] = [this.find(union, src), this.find(union, dst)];

      const hasCycle = x === y;
      if (hasCycle) return false;

      this.compress(union, x, y);
    }

    const isValid = edges.length === n - 1;
    return isValid;
  }

  compress = (union, i, head) => (union[i] = head);

  find = (union, i, num = union[i]) => {
    const isEmpty = num === -1;
    if (isEmpty) return i;

    const head = this.find(union, num);

    this.compress(union, i, head);

    return union[i];
  };

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(...input));
        else this.method(...input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(...input));
        else this.reference(...input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  [
    5,
    [
      [0, 1],
      [0, 2],
      [0, 3],
      [1, 4],
    ],
  ],
  // Additional
  [
    7,
    [
      [0, 1],
      [0, 2],
      [3, 5],
      [5, 6],
      [1, 4],
    ],
  ],
];
test.quantify(testCases);
